### ISSUE TITLE
Json serialize when use custom nullable JsonConverter always thrown AccessViolationException

### ISSUE THREAD

**Issue by Cricle** [dotnet/runtime#55135]
### Description

When use nullable JsonConverter to custom serialize double? and double, always thrown  AccessViolationException.

### Configuration

|Item|Version|
|:-:|:-:|
|.NET Runtime|5.0.301|
|TargetFramework|5.0|
|System.Text.Json|5.0.1|

### Regression?

Never normal

### Other information

<img width="766" alt="1" src="https://user-images.githubusercontent.com/26201412/124371229-3962d580-dcb2-11eb-8a25-5d8187471a23.PNG">

**Comment by eiriktsarpalis** [comment]
I can reproduce the issue in both .NET 5 and 6. Transcribing the minimal repro:
```csharp
using System;
using System.Text.Json;
using System.Text.Json.Serialization;

JsonSerializer.Serialize(new NullableD(), new JsonSerializerOptions { Converters = { new NullableDouble() } });

public class NullableD
{
    public double B { get; set; }
}

public class NullableDouble : JsonConverter<double?>
{
    public override bool CanConvert(Type typeToConvert) => typeToConvert == typeof(double);

    public override double? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) => 0;

    public override void Write(Utf8JsonWriter writer, double? value, JsonSerializerOptions options) => writer.WriteNullValue();
}
```
which throws
```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at DynamicClass.BGetter(System.Object)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1[[System.Nullable`1[[System.Double, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].GetMemberAndWriteJson(System.Object, System.Text.Json.WriteStack ByRef, System.Text.Json.Utf8JsonWriter)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].OnTryWrite(System.Text.Json.Utf8JsonWriter, System.__Canon, System.Text.Json.JsonSerializerOptions, System.Text.Json.WriteStack ByRef)
   at System.Text.Json.Serialization.JsonConverter`1[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].TryWrite(System.Text.Json.Utf8JsonWriter, System.__Canon ByRef, System.Text.Json.JsonSerializerOptions, System.Text.Json.WriteStack ByRef)
   at System.Text.Json.Serialization.JsonConverter`1[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].WriteCore(System.Text.Json.Utf8JsonWriter, System.__Canon ByRef, System.Text.Json.JsonSerializerOptions, System.Text.Json.WriteStack ByRef)
   at System.Text.Json.JsonSerializer.WriteCore[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.Text.Json.Serialization.JsonConverter, System.Text.Json.Utf8JsonWriter, System.__Canon ByRef, System.Text.Json.JsonSerializerOptions, System.Text.Json.WriteStack ByRef)
   at System.Text.Json.JsonSerializer.WriteUsingMetadata[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.Text.Json.Utf8JsonWriter, System.__Canon ByRef, System.Text.Json.Serialization.Metadata.JsonTypeInfo)
   at System.Text.Json.JsonSerializer.WriteUsingMetadata[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.__Canon ByRef, System.Text.Json.Serialization.Metadata.JsonTypeInfo)
   at System.Text.Json.JsonSerializer.Write[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.__Canon ByRef, System.Type, System.Text.Json.JsonSerializerOptions)
   at System.Text.Json.JsonSerializer.Serialize[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.__Canon, System.Text.Json.JsonSerializerOptions)
   at <Program>$.<Main>$(System.String[])
Segmentation fault
```
I believe the issue is caused by invalid IL emitted by the Reflection.Emit member accessor. The converter implementation is incorrect, however we should detecting this error earlier and fail gracefully with a helpful error message.

**Comment by wzchua** [comment]
Should this be an `||` here?
https://github.com/dotnet/runtime/blob/90b8ddd7a6b28e57aebed7d478547ce07b9b60a6/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs#L253-L257

**Comment by eiriktsarpalis** [comment]
> Should this be an || here?

I think that this check is intentional, apparently it is meant to support "covariant" and "contravariant" converters and will only fail if there is no subtype relationship between the two types. Here are some tests that validate this:

https://github.com/dotnet/runtime/blob/a8dacd5577b4da88555f342abc8d5a3f219edca0/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/CustomConverterTests/CustomConverterTests.BadConverters.cs#L46-L57

I find that this mode is flawed since serialization is contravariant and deserialization is covariant, meaning that covariant serialization and contravariant deserialization will necessarily manifest as various kinds of runtime type errors. On top of that, much of the serialization infrastructure is built on the assumption that `JsonConverter<T>` is invariant, see for example:

* [IEnumerableDefaultConverter](https://github.com/dotnet/runtime/blob/a036594cb3766f70db9f5c6e7681454e6894da50/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableDefaultConverter.cs#L22)
* [DictionaryDefaultConverter](https://github.com/dotnet/runtime/blob/a036594cb3766f70db9f5c6e7681454e6894da50/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs#L43)
* [JsonPropertyInfoOfT](https://github.com/dotnet/runtime/blob/a036594cb3766f70db9f5c6e7681454e6894da50/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfoOfT.cs#L225)
* [ObjectWithParameterizedConstructorConverter](https://github.com/dotnet/runtime/blob/a036594cb3766f70db9f5c6e7681454e6894da50/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.Small.cs#L66)

and therefore any "contraviariant"/"covariant" converter will necessarily manifest as an `InvalidCastException` when composed with collection converters or constructor arguments.

Ultimately, that is the root cause for the reported issue: `JsonPropertyInfo` instances are [constructed based on the converter type](https://github.com/dotnet/runtime/blob/a8dacd5577b4da88555f342abc8d5a3f219edca0/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L59) and not the `PropertyInfo` type. This in turn results in us [emitting dynamic method accessors that contain type errors](https://github.com/dotnet/runtime/blob/a8dacd5577b4da88555f342abc8d5a3f219edca0/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfoOfT.cs#L68), which causes the `AccessViolationException`.

One was of addressing this issue as well as #46522 is to force converter invariance everywhere, namely changing the check in [`GetConverterInternal`](https://github.com/dotnet/runtime/blob/90b8ddd7a6b28e57aebed7d478547ce07b9b60a6/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs#L253-L257) to say
```csharp

            if (converter.TypeToConvert != typeToConvert)
            {
                ThrowHelper.ThrowInvalidOperationException_SerializationConverterNotCompatible(converter.GetType(), typeToConvert);
            }
```
This will ensure consistent behavior across the board (regardless of whether the converter is used as a root-level object or collection element) and any type mismatches will manifest as the same user-friendly error messages occurring early in the serialization process. Perhaps something to consider for .NET 7.

@layomia @steveharter @ericstj thoughts?

**Comment by steveharter** [comment]
We should not be causing AccessViolation. Ideally we improve the error detection upfront. If that is not feasible, we could add some `castclass` opcodes in the IL generation that will throw `InvalidCastException`, but that will have a small impact on runtime perf.

The converter model was not intended to be covariant - the JsonConverterFactory should be used to instantiate the correct converter.

**Comment by steveharter** [comment]
This could be moved to 7.0; I don't believe the semantics changed in 6.0.

**Comment by eiriktsarpalis** [comment]
I already have a PR open that fixes this: #56577
